source/sn_coap_protocol.c

```diff
@@ -1207,67 +1212,91 @@ static void sn_coap_protocol_linked_list_blockwise_payload_store(struct coap_s *
+    coap_blockwise_payload_s *stored_blockwise_payload_ptr = sn_coap_protocol_linked_list_blockwise_search(handle, addr_ptr, token_ptr, token_len);
 
-    if (stored_blockwise_payload_ptr == NULL) {
-        tr_error("sn_coap_protocol_linked_list_blockwise_payload_store - failed to allocate blockwise!");
-        return;
-    }
+    if (stored_blockwise_payload_ptr && stored_blockwise_payload_ptr->use_size1) {
+        memcpy(stored_blockwise_payload_ptr->payload_ptr + (block_number * handle->sn_coap_block_data_size), payload_ptr, payload_len);
+    } else if (stored_blockwise_payload_ptr) {
+        uint16_t new_len = stored_blockwise_payload_ptr->payload_len + payload_len;
+        tr_debug("sn_coap_protocol_linked_list_blockwise_payload_store - reallocate from %d to %d", stored_blockwise_payload_ptr->payload_len, new_len);
 
+        stored_blockwise_payload_ptr->payload_ptr = realloc(stored_blockwise_payload_ptr->payload_ptr, new_len);
+        if (stored_blockwise_payload_ptr->payload_ptr == NULL) {
+            tr_error("sn_coap_protocol_linked_list_blockwise_payload_store - failed to reallocate payload!");
+            handle->sn_coap_protocol_free(stored_blockwise_payload_ptr->payload_ptr);
+            return;
+        }
+        memcpy(stored_blockwise_payload_ptr->payload_ptr + stored_blockwise_payload_ptr->payload_len, payload_ptr, payload_len);
+        stored_blockwise_payload_ptr->payload_len = new_len;
+    } else {
+        stored_blockwise_payload_ptr = NULL;
+        stored_blockwise_payload_ptr = handle->sn_coap_protocol_malloc(sizeof(coap_blockwise_payload_s));
 
-    /* Allocate memory for stored Payload's data */
-    stored_blockwise_payload_ptr->payload_ptr = sn_coap_protocol_malloc_copy(handle, stored_payload_ptr, stored_payload_len);
+        if (stored_blockwise_payload_ptr == NULL) {
+            tr_error("sn_coap_protocol_linked_list_blockwise_payload_store - failed to allocate blockwise!");
+            return;
+        }
 
-    if (stored_blockwise_payload_ptr->payload_ptr == NULL) {
-        tr_error("sn_coap_protocol_linked_list_blockwise_payload_store - failed to allocate payload!");
-        handle->sn_coap_protocol_free(stored_blockwise_payload_ptr);
-        return;
-    }
+        // If size1 option present then we can allocate the whole payload before hand.
+        stored_blockwise_payload_ptr->use_size1 = size1 ? 1 : 0;
 
-    /* Allocate memory for stored Payload's address */
-    stored_blockwise_payload_ptr->addr_ptr = sn_coap_protocol_malloc_copy(handle, addr_ptr->addr_ptr, addr_ptr->addr_len);
+        /* Allocate memory for stored Payload's data */
+        if (stored_blockwise_payload_ptr->use_size1) {
+            stored_blockwise_payload_ptr->payload_ptr = handle->sn_coap_protocol_malloc(size1);
+            if (stored_blockwise_payload_ptr->payload_ptr) {
+                memset(stored_blockwise_payload_ptr->payload_ptr, 0x0, size1);
```

Comment:
What is this memset() for? Now the valgrind will not notice if the payload was not filled before sending.

Reply:
Fixed.